wp-env: granular volume mappings

noahtallen opened the pull request:
## Description
Adds support for granular WordPress directory mappings. For example, you can do the following in `.wp-env.json`:

```json
{
  "mappings": {
    "wp-content/plugins": "./path/to/lots-of-plugins"
  }
}
```

For the time being, I have only permitted local directory sources as values here. If folks want, though, it would be very easy to add git/zip source support to this as well.

## How has this been tested?
Tested running wp-env start locally with these changes. I made sure that the mapped directory was mounted in WordPress, and that the plugins I expected to see also showed up in the plugins list. I also tested with themes.


## Types of changes
New feature.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

noahtallen committed:
Add mappings entry to config object

noahtallen committed:
Mount each "mapping" entry in Docker image

- Moved test plugins to .wp-env.json

noahtallen committed:
Add documentation

noahtallen committed:
Add tests for mapping

- include new test file for build-docker-compose-config

epiqueras commented on .wp-env.json:
```diff
@@ -1,4 +1,7 @@
 {
 	"core": "WordPress/WordPress",
-	"plugins": [ "." ]
+	"plugins": [ ".", "./packages/e2e-tests/plugins" ],
```
Why are the test plugins here?

epiqueras replied:
So they're activated?

epiqueras replied:
We would need to have the `env` separation then, otherwise, this would make the development environment very hard to use.

noahtallen replied:
They aren't activated because `env` doesn't see into the directory of plugins. But yeah, the `env` separation is the ultimate goal here.

noahtallen replied:
lol, `env`, as in `wp-env`, doesn't see into the directory. so it can't call plugin activate on any of the names there.

epiqueras replied:
```js
// Activate all plugins.
	for ( const pluginSource of config.pluginSources ) {
		await dockerCompose.run(
			environment === 'development' ? 'cli' : 'tests-cli',
			`wp plugin activate ${ pluginSource.basename }`,
			options
		);
	}
```

That still runs, right? Will it just fail silently?

noahtallen replied:
> Will it just fail silently

yes. I can just add them as a mapping, though, to avoid any confusion.

noahtallen replied:
done. also fixed a few other things and added a mu-plugins example

noahtallen committed:
Support any source in mapping config

- Update changelog and readme to reference mu-plugins
- support git/zip sources
- add more examples to readme
- fix readme type

noahtallen commented:
interesting, I'm getting an error trying to run `wp-env start` that matches this issue: https://github.com/docker/compose/issues/3633

noahtallen commented:
Ah never mind that. It was because Docker wanted to install an update.

epiqueras commented on packages/env/README.md:
````diff
@@ -323,6 +324,35 @@ This is useful for integration testing: that is, testing how old versions of Wor
+```json
+{
+	"plugins": [ "." ],
+	"themes": [ "WordPress/theme-experiments" ],
````
This should show a specific theme.

epiqueras commented on packages/env/README.md:
````diff
@@ -323,6 +324,35 @@ This is useful for integration testing: that is, testing how old versions of Wor
+}
+```
+
+#### Avoid activating plugins on the instance
````
We should mention the same applies to themes.

epiqueras commented on packages/env/test/config.js:
```diff
@@ -219,6 +219,92 @@ describe( 'readConfig', () => {
+		}
+	} );
+
+	it( 'excludes undefined sources', async () => {
```
Why is that needed? Shouldn't we throw instead?

epiqueras commented on packages/env/package.json:
```diff
@@ -39,6 +39,7 @@
 		"got": "^10.7.0",
 		"inquirer": "^7.1.0",
 		"js-yaml": "^3.13.1",
+		"lodash": "^4.17.15",
```
We can use `Object.keys`, `Object.values`, and `Object.entries` to avoid using Lodash.

noahtallen committed:
Clarify readme

noahtallen committed:
Remove lodash, validate mapping keys

noahtallen commented:
Thanks for the comments, I've resolved each of those now :) I'll merge once I get a ✅ from travis

aduth commented:
Is there some way that mappings can differentiate between the test and non-test environment? For example, the current `mu-plugins` mapping for Gutenberg should arguably only apply for the test environment. The fact that they don't has caused some confusion in the past (e.g. https://github.com/WordPress/gutenberg/pull/14245#issuecomment-470029910, https://github.com/WordPress/gutenberg/pull/18761#issuecomment-615209811).

I could possibly see an argument that the mappings should be kept simple and apply universally, though it's the same sort of argument I think would justify that "test" and "non-test" sites should just be multiple environments, not a first-class distinction.

I'm also open to alternative approaches. Curious if there are any thoughts on this.

- Avoid using `mu-plugins` for end-to-end tests and just use regular plugins that are activated in the test environment. The primary downside here is that this causes some delay in test execution time.
- Somehow detect environment from within the implementation of the `mu-plugin` as recognizing to be the test or non-test environment. Possibly via the availability of some environment variable exposed from `wp-env` ?

noahtallen commented:
> For example, the current mu-plugins mapping for Gutenberg should arguably only apply for the test environment.

Yeah, this has been an issue since wp-env was created. Before this PR, is was hidden in the code.

> Is there some way that mappings can differentiate between the test and non-test environment?

Not yet. we started talking about this here: https://github.com/WordPress/gutenberg/pull/21229#issuecomment-620137794

I have a WIP branch which allows for different configurations in each environment, which would also solve #22514. I haven't made a lot of progress on it lately, but I can bring it back up since it comes up a lot.

noahtallen commented:
@aduth WIP pull request here if you'd like to follow along: https://github.com/WordPress/gutenberg/pull/22568
